client/driver/docker_progress.go

```diff
@@ -0,0 +1,242 @@
+	totalBytes   int64
+}
+
+type layerProgressStatus int
+
+const (
+	layerProgressStatusUnknown layerProgressStatus = iota
+	layerProgressStatusStarting
+	layerProgressStatusWaiting
+	layerProgressStatusDownloading
+	layerProgressStatusVerifying
+	layerProgressStatusDownloaded
+	layerProgressStatusExtracting
+	layerProgressStatusComplete
+	layerProgressStatusExists
+)
+
+func lpsFromString(status string) layerProgressStatus {
+	switch status {
+	case "Pulling fs layer":
+		return layerProgressStatusStarting
+	case "Waiting":
+		return layerProgressStatusWaiting
+	case "Downloading":
+		return layerProgressStatusDownloading
+	case "Verifying Checksum":
+		return layerProgressStatusVerifying
+	case "Download complete":
+		return layerProgressStatusDownloaded
+	case "Extracting":
+		return layerProgressStatusExtracting
+	case "Pull complete":
+		return layerProgressStatusComplete
+	case "Already exists":
+		return layerProgressStatusExists
+	default:
+		return layerProgressStatusUnknown
+	}
+}
+
+type imageProgress struct {
+	sync.RWMutex
+	lastMessage *jsonmessage.JSONMessage
+	timestamp   time.Time
+	layers      map[string]*layerProgress
+	pullStart   time.Time
+}
+
+func (p *imageProgress) get() (string, time.Time) {
+	p.RLock()
+	defer p.RUnlock()
+
+	if p.lastMessage == nil {
+		return "No progress", p.timestamp
+	}
+
+	var pulled, pulling int
+	for _, l := range p.layers {
+		if l.status == layerProgressStatusDownloading {
+			pulling++
+		} else if l.status > layerProgressStatusVerifying {
+			pulled++
+		}
+	}
+
+	elapsed := time.Now().Sub(p.pullStart)
+	cur := p.currentBytes()
+	total := p.totalBytes()
+	var est int64
+	if cur != 0 {
+		est = (elapsed.Nanoseconds() / cur * total) - elapsed.Nanoseconds()
+	}
+
+	return fmt.Sprintf("Pulled %d/%d (%d/%dMB) pulling %d layers - est %.1fs remaining",
+		pulled, len(p.layers), cur/1000/1000, total/1000/1000, pulling,
+		time.Duration(est).Seconds()), p.timestamp
+}
+
+func (p *imageProgress) set(msg *jsonmessage.JSONMessage) {
+	p.Lock()
+	defer p.Unlock()
+
+	p.lastMessage = msg
+	p.timestamp = time.Now()
+
+	lps := lpsFromString(msg.Status)
+	if lps == layerProgressStatusUnknown {
+		return
+	}
+
+	layer, ok := p.layers[msg.ID]
+	if !ok {
+		layer = &layerProgress{id: msg.ID}
+		p.layers[msg.ID] = layer
+	}
+	layer.status = lps
+	if msg.Progress != nil && lps == layerProgressStatusDownloading {
+		layer.currentBytes = msg.Progress.Current
+		layer.totalBytes = msg.Progress.Total
+	} else if lps == layerProgressStatusDownloaded {
+		layer.currentBytes = layer.totalBytes
+	}
+}
+
+func (p *imageProgress) currentBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.currentBytes
+	}
+	return b
+}
+
+func (p *imageProgress) totalBytes() int64 {
+	var b int64
+	for _, l := range p.layers {
+		b += l.totalBytes
+	}
+	return b
+}
+
+type progressReporterFunc func(image string, msg string, timestamp time.Time, pullStart time.Time)
+
+type imageProgressManager struct {
+	*imageProgress
```

Comment:
Not sure if I'm a fan of this embedding. It causes `imageProgressManager` to have RWMutex's methods despite those methods only guarding fields on `imageProgress`. Perhaps just give it a named field to make it clear for uses in `imageProgressManager` which structs "owns" which fields.

Not a big deal. I do like the overall approach of bundling up the fields that need guarded by a mutex with a little oneshot struct.